src/Neo/Extensions/NeoSystemExtensions.cs

```diff
@@ -20,6 +21,44 @@ namespace Neo
     /// </summary>
     public static class NeoSystemExtensions
     {
+        /// <summary>
+        /// Gets the block generation time based on the current state of the blockchain.
+        /// </summary>
+        /// <param name="system">The NeoSystem instance.</param>
+        /// <returns>The block generation time as a TimeSpan.</returns>
+        public static TimeSpan GetBlockGenTime(this NeoSystem system)
+        {
+            return system.StoreView.GetBlockGenTime(system.Settings);
+        }
+
+        /// <summary>
+        /// Gets the block generation time based on the current state of the blockchain.
+        /// </summary>
+        /// <param name="snapshot">The snapshot of the store.</param>
+        /// <param name="settings">The protocol settings.</param>
+        /// <returns>The block generation time as a TimeSpan.</returns>
+        public static TimeSpan GetBlockGenTime(this IReadOnlyStore snapshot, ProtocolSettings settings)
+        {
+            try
+            {
+                // Get the current block height from the blockchain
+                var index = NativeContract.Ledger.CurrentIndex(snapshot);
+
+                // Before the Echidna hardfork, use the protocol settings
+                if (!settings.IsHardforkEnabled(Hardfork.HF_Echidna, index))
+                    return TimeSpan.FromMilliseconds(settings.MillisecondsPerBlock);
+
+                // After the Echidna hardfork, get the current block time from the Policy contract
+                var milliseconds = NativeContract.Policy.GetMSPerBlock(snapshot);
+                return TimeSpan.FromMilliseconds(milliseconds);
+            }
+            catch (KeyNotFoundException)
+            {
+                // At the height of 0, the key not yet exists in the storage
+                return TimeSpan.FromMilliseconds(settings.MillisecondsPerBlock);
```

Comment:
When is the key added to the storage? I am thinking about when we set HF_Echidna to 0, then we may have problems

Reply:
> When is the key added to the storage? 

At the block of HF_Echidna activation.

> I am thinking about when we set HF_Echidna to 0, then we may have problems

Why? I'll check it on privnet, but in general I don't think there will be any problem, because this value is available right after genesis block's OnPersist trigger is processed (if we're talking about `HF_Echidna: 0` setting).